content/blog/pitfalls-of-safe-rust/index.md

```diff
@@ -0,0 +1,687 @@
+use serde::{Deserialize, Deserializer};
+
+#[derive(Serialize)]
+struct UserCredentials {
+    username: Username,
+    #[serde(skip_serializing)]  // Never serialize passwords
+    password: Password,
+}
+
+impl<'de> Deserialize<'de> for UserCredentials {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        struct Raw {
+            username: String,
+            password: String,
+        }
+
+        let raw = Raw::deserialize(deserializer)?;
+        
+        // Validate during deserialization
+        // Throw an error if the username or password is invalid
+        let username = Username::new(&raw.username)
+            .map_err(serde::de::Error::custom)?;
+        let password = Password::new(&raw.password)
+            .map_err(serde::de::Error::custom)?;
+
+        Ok(UserCredentials { username, password })
+    }
+}
+```
+
+## Protect Against Time-of-Check to Time-of-Use (TOCTOU)
+
+This is a more advanced topic, but it's important to be aware of it.
+TOCTOU is a class of software bugs caused by attackers exploiting the gap between the time a condition is checked and the time it is used. 
+Simply put, an attacker could rug-pull the value from under your feet
+and you wouldn't notice because it happens exactly between your check and use.
+
+```rust
+// DON'T: Check and use in separate operations
+fn process_file(path: &Path) -> Result<(), io::Error> {
+    if path.exists() {  // Check
+        let content = fs::read(path)?;  // Use - file might have changed!
+        // Process content
+    }
+    Ok(())
+}
+```
+
+Rust supports atomic file operations to prevent this kind of attack.
+
+```rust
+// DO: Use atomic operations when possible
+fn process_file(path: &Path) -> Result<(), io::Error> {
+    let file = fs::OpenOptions::new()
+        .read(true)
+        .write(true)
+        .create_new(true)  // Atomic operation
+        .open(path)?;
+    // Process file
+    Ok(())
+}
+```
+
+It's a bit more verbose, but it's safer.
+I think more Rustaceans should know about this pattern.
+
+
+## Use Constant-Time Comparison for Sensitive Data
+
+Timing attacks are a nifty way to extract information from your application.
+The idea is that the time it takes to compare two values can leak information about them.
+For example, the time it takes to compare two strings can reveal how many characters are correct.
+Therefore, for production code, be careful with regular equality checks when handling sensitive data like passwords.
+
+```rust
+// DON'T: Use regular equality for sensitive comparisons
+fn verify_password(stored: &[u8], provided: &[u8]) -> bool {
+    stored == provided  // Vulnerable to timing attacks!
+}
+
+// DO: Use constant-time comparison
+use subtle::{ConstantTimeEq, Choice};
+
+fn verify_password(stored: &[u8], provided: &[u8]) -> bool {
+    if stored.len() != provided.len() {
```

Comment:
I don't know much about this area but why check the length first? Isn't that another timing leak that could be used to deduce the length?

Reply:
Ah, gotcha. My performance optimization backfired! Let me remove that again. Thanks for pointing that out.